tracker/peerstore/local.go

```diff
@@ -0,0 +1,161 @@
+}
+
+type peerGroup struct {
+	mu            sync.Mutex
+	peers         map[core.PeerID]*peerEntry
+	lastExpiresAt time.Time
+	deleted       bool
+}
+
+type peerEntry struct {
+	ip        string
+	port      int
+	complete  bool
+	expiresAt time.Time
+}
+
+type cleanupTask struct {
+	store *LocalStore
+}
+
+func (t *cleanupTask) Run() {
+	t.store.runCleanup()
+}
+
+// NewLocalStore creates a new LocalStore.
+func NewLocalStore(config LocalConfig, clk clock.Clock) *LocalStore {
+	config.applyDefaults()
+	s := &LocalStore{
+		config:     config,
+		clk:        clk,
+		peerGroups: make(map[core.InfoHash]*peerGroup),
+	}
+	s.cleanup = dedup.NewIntervalTrap(_peerGroupCleanupInterval, clk, &cleanupTask{s})
+	return s
+}
+
+// GetPeers implements Store.
+func (s *LocalStore) GetPeers(h core.InfoHash, n int) ([]*core.PeerInfo, error) {
+	if n <= 0 {
+		// Simpler for below logic to assume positive n.
+		return nil, nil
+	}
+
+	s.mu.RLock()
+	g, ok := s.peerGroups[h]
+	s.mu.RUnlock()
+	if !ok {
+		return nil, nil
+	}
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	now := s.clk.Now()
+	var result []*core.PeerInfo
+
+	// We rely on random map iteration to pick n random peers.
+	for id, p := range g.peers {
+		if now.After(p.expiresAt) {
+			// Clean up any expired peers we run into.
+			delete(g.peers, id)
+			continue
+		}
+		result = append(result, core.NewPeerInfo(id, p.ip, p.port, false /* origin */, p.complete))
+		if len(result) == n {
+			break
+		}
+	}
+	return result, nil
+}
+
+// UpdatePeer implements Store.
+func (s *LocalStore) UpdatePeer(h core.InfoHash, p *core.PeerInfo) error {
+	s.cleanup.Trap()
+
+	g := s.getOrInitLockedPeerGroup(h)
+	defer g.mu.Unlock()
+
+	expiresAt := s.clk.Now().Add(s.config.TTL)
+
+	g.peers[p.PeerID] = &peerEntry{
+		ip:        p.IP,
+		port:      p.Port,
+		complete:  p.Complete,
+		expiresAt: expiresAt,
+	}
+
+	// Allows runCleanup to quickly determine when the last peerEntry expires.
+	g.lastExpiresAt = expiresAt
+
+	return nil
+}
+
+func (s *LocalStore) getOrInitLockedPeerGroup(h core.InfoHash) *peerGroup {
+	// We must take care to handle a race condition against runCleanup. Consider
+	// two goroutines, A and B, where A executes getOrInitLockedPeerGroup and B
+	// executes runCleanup:
+	//
+	// A: locks s.mu, reads g from s.peerGroups, unlocks s.mu
+	// B: locks s.mu, locks g.mu, deletes g from s.peerGroups, unlocks g.mu
+	// A: locks g.mu
+	//
+	// At this point, A is holding onto a peerGroup reference which has been
+	// deleted from the peerGroups map, and thus has no choice but to attempt to
+	// reload a new peerGroup. Since the cleanup interval is quite large, it is
+	// *extremeley* unlikely this for-loop will execute more than twice.
+	for {
+		s.mu.Lock()
+		g, ok := s.peerGroups[h]
+		if !ok {
+			g = &peerGroup{
+				peers:         make(map[core.PeerID]*peerEntry),
+				lastExpiresAt: s.clk.Now().Add(s.config.TTL),
+			}
+			s.peerGroups[h] = g
+		}
+		s.mu.Unlock()
+
+		g.mu.Lock()
```

Comment:
I did consider it -- in fact, that's how one would normally implement a two-level lock.

With Kraken, I usually like to consider the max number of readers to be 10k, which corresponds to an entire cluster downloading the same hash simultaneously. This means you could have 10k GetPeers goroutines competing to lock a single peerGroup mutex; in other words, I'd expect peerGroup mutex contention to be quite high.

What this means is if an UpdatePeer goroutine must hold the top-level mutex while waiting for a peerGroup mutex, it needs to compete against 10k GetPeers goroutines for the peerGroup mutex. And while this is happening, _every_ hash (i.e. peerGroup) in the map is inaccessible because the top-level mutex is locked. We've now limited throughput of all hashes to the throughput of the most contentious hash.

Reply:
> or alternatively, having readers acquire a read lock on the peergroup by default and only attempting to acquire a writelock if the entry is indeed stale.

This is definitely a valid optimization. I don't expect GetPeers will run into expired entries often, given that the default TTL is 5 hours. So for the first 5 hours, you'll never actually need a write lock. Let me look into this.

Reply:
both your comments sgtm